api/commands/run.ts

```diff
@@ -61,134 +83,223 @@ export class Runner {
+              task: this.graph.actions.find(action => action.name === actionResult.name).tasks[i]
+            };
+            if (!!taskResult.errorMessage) {
+              executedTask.error = taskResult.errorMessage;
+            }
+            return executedTask;
+          }),
+          executionTime: actionResult.timing.endTimeMillis.subtract(
+            actionResult.timing.startTimeMillis
+          ),
+          deprecatedOk: isSuccessfulAction(actionResult)
+        }))
+    };
+    if (!!this.runResult.status) {
+      executedGraph.ok = this.runResult.status === dataform.RunResult.ExecutionStatus.SUCCESSFUL;
+    }
+    return executedGraph;
   }
 
   private async executeGraph() {
+    const timer = Timer.start();
+
+    this.runResult.status = dataform.RunResult.ExecutionStatus.RUNNING;
+    this.runResult.timing = timer.current();
+    await this.triggerChange();
+
+    // Work out all the schemas we are going to need to create first.
+    const uniqueSchemas: { [schema: string]: boolean } = {};
+    this.graph.actions
+      .filter(action => !!action.target)
+      .map(action => action.target.schema)
+      .filter(schema => !!schema)
+      .forEach(schema => (uniqueSchemas[schema] = true));
+
+    // Wait for all schemas to be created.
+    await Promise.all(Object.keys(uniqueSchemas).map(schema => this.adapter.prepareSchema(schema)));
+
     // Recursively execute all actions as they become executable.
     await this.executeAllActionsReadyForExecution();
 
-    let ok = true;
-    this.result.actions.forEach(action => {
-      ok = ok && isSuccessfulAction(action);
-    });
-    this.result.ok = ok;
+    this.runResult.timing = timer.end();
+
+    this.runResult.status = dataform.RunResult.ExecutionStatus.SUCCESSFUL;
+    if (
+      this.runResult.actions.filter(
+        action => action.status === dataform.ActionResult.ExecutionStatus.CANCELLED
+      ).length > 0
+    ) {
+      this.runResult.status = dataform.RunResult.ExecutionStatus.CANCELLED;
+    }
+    if (
+      this.runResult.actions.filter(
+        action => action.status === dataform.ActionResult.ExecutionStatus.FAILED
+      ).length > 0
+    ) {
+      this.runResult.status = dataform.RunResult.ExecutionStatus.FAILED;
+    }
+
+    return this.runResult;
   }
 
   private async executeAllActionsReadyForExecution() {
-    // If the run has been cancelled, skip all pending actions.
+    // If the run has been cancelled, cancel all pending actions.
     if (this.cancelled) {
       const allPendingActions = this.pendingActions;
       this.pendingActions = [];
-      await Promise.all(
-        allPendingActions.map(async pendingAction =>
-          this.onActionExecutionComplete(toSkippedAction(pendingAction))
-        )
+      allPendingActions.forEach(pendingAction =>
+        this.runResult.actions.push({
+          name: pendingAction.name,
+          status: dataform.ActionResult.ExecutionStatus.CANCELLED,
```

Comment:
IMO these are skipped, not cancelled. Or rather, I interpreted "cancelled" to mean "interrupted" actions/tasks. I think they is value in knowing which ones were killed, as opposed to which ones never started because the run was cancelled